tests/satosa/micro_services/test_custom_routing.py

```diff
@@ -156,3 +158,77 @@ def test_missing_target_entity_id_from_context(self, context):
         req = InternalData(requester="test_requester")
         with pytest.raises(SATOSAError):
             decide_service.process(context, req)
+
+
+class TestDecideBackendByTarget:
+    rules = {
+        'default_backend': 'Saml2',
+        'endpoint_paths': ['.*/disco'],
+        'target_mapping': {'http://idpspid.testunical.it:8088': 'spidSaml2'}
+    }
+
+    def create_decide_service(self, rules):
+        decide_service = DecideBackendByTarget(
+                config=rules,
+                name="test_decide_service",
+                base_url="https://satosa.example.com"
+        )
+        decide_service.next = lambda ctx, data: data
+        return decide_service
+
+
+    def test_missing_state(self, target_context):
+        decide_service = self.create_decide_service(self.rules)
+        target_context.request = {
+            'entityID': 'http://idpspid.testunical.it:8088',
+        }
+        req = InternalData(requester="test_requester")
+        req.requester = "somebody else"
+        assert decide_service.process(target_context, req)
+
+        with pytest.raises(SATOSAStateError):
+            decide_service.backend_by_entityid(target_context)
+
+
+    def test_missing_entityid(self, target_context):
+        decide_service = self.create_decide_service(self.rules)
+        target_context.request = {
+            # 'entityID': None,
+        }
+        target_context.state['ROUTER'] = 'Saml2'
+
+        req = InternalData(requester="test_requester")
+        assert decide_service.process(target_context, req)
+
+        with pytest.raises(CustomRoutingError):
+            decide_service.backend_by_entityid(target_context)
+
+
+    def test_unmatching_target(self, target_context):
+        """
+            It would rely on the default backend
+        """
+        decide_service = self.create_decide_service(self.rules)
+        target_context.request = {
+            'entityID': 'unknow-entity-id',
+        }
+        target_context.state['ROUTER'] = 'Saml2'
+
+        req = InternalData(requester="test_requester")
+        assert decide_service.process(target_context, req)
+
+        res = decide_service.backend_by_entityid(target_context)
+        assert isinstance(res, InternalData)
```

Comment:
you want to ensure that the target_backend is the expected one.
On `create_decide_service`, do
```diff
-        decide_service.next = lambda ctx, data: data
+        decide_service.next = lambda ctx, data: ctx, data
```
then
```suggestion
        expected_backend = rules['default_backend']
        res = decide_service.backend_by_entityid(target_context)
        assert ctx.get_decoration(Context.KEY_TARGET_ENTITYID) == expected_backend
```